File: dotnet/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example2/Folder.cs
namespace CompositeDesignPattern.Example2;

public class Folder : IFileSystemItem
{
	private List<IFileSystemItem> _items = new();

	public Folder(string name, string path) : base(name, path)
	{
	}

	public override double GetSize()
	{
		return _items.Sum(item => item.GetSize());
	}

	public void AddItem(IFileSystemItem item)
	{
		_items.Add(item);
	}

	public void RemoveItem(IFileSystemItem item)
	{
		_items.Remove(item);
	}

	public void DisplayItems()
	{
		foreach (var item in _items)
		{
			if (item is Folder folder)
			{
				Console.WriteLine($"Folder: {folder.Name}, Path: {folder.Path}, Created: {folder.CreatedDate}, Size: {folder.GetSize()} bytes");
				folder.DisplayItems(); // Recursive call to display sub-items
			}
			else if (item is File file)
			{
				Console.WriteLine($"File: {file.Name}, Path: {file.Path}, Created: {file.CreatedDate}, Size: {file.GetSize()} bytes");
			}
		}
	}
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example2/src/Folder.ts
import { FileSystemItem } from './FileSystemItem';
import { File } from './File';

export class Folder extends FileSystemItem {
  private readonly items: FileSystemItem[] = [];

  constructor(name: string, path: string) { super(name, path); }

  getSize(): number { return this.items.reduce((sum, i) => sum + i.getSize(), 0); }

  addItem(item: FileSystemItem): void { this.items.push(item); }
  removeItem(item: FileSystemItem): void {
    const idx = this.items.indexOf(item);
    if (idx !== -1) this.items.splice(idx, 1);
  }

  displayItems(): void {
    for (const item of this.items) {
      if (item instanceof Folder) {
        console.log(`Folder: ${item.name}, Path: ${item.path}, Created: ${item.createdDate.toISOString()}, Size: ${item.getSize()} bytes`);
        item.displayItems();
      } else if (item instanceof File) {
        console.log(`File: ${item.name}, Path: ${item.path}, Created: ${item.createdDate.toISOString()}, Size: ${item.getSize()} bytes`);
      }
    }
  }
}
